low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -22,28 +22,227 @@
+        return node, inst
+
+    def get_data_template(self, uid, is_fru: bool):
+        return {
+            "uid": uid,
+            "fru": str(is_fru).lower(),
+            "last_updated": int(time.time()),
+            "health": {
+                "status": "",
+                "description": "",
+                "recommendation": "",
+                "specifics": []
+            }
+        }
+
+    def set_health_data(self, obj: dict, status, description=None,
+                        recommendation=None):
+        good_state = (status == "OK")
+        if not description:
+            description = "%s is %s in good state" % (
+                            obj.get("uid"),
+                            '' if good_state else 'not')
+        if not recommendation:
+            recommendation = 'None' if good_state\
+                             else "Contact Seagate Support."
+
+        obj["health"].update({
+            "status": status,
+            "description": description,
+            "recommendation": recommendation
+        })
+
+    def get_sas_hba_info(self):
+        return self.get_sas_ports_info("sas_hba")
+
+    def get_sas_ports_info(self, leaf_node="sas_ports"):
+        sas_hba_data = []
+        health_data = []
+        for sas_rid in self.SAS_RESOURCE_ID:
+            r_data = self.get_data_template(sas_rid, False)
+            sysfs_instance = ToolFactory().get_instance('sysfs')
+            sysfs_instance.initialize()
+            sas_ports_dict = sysfs_instance.get_phy_negotiated_link_rate()
+            sas_ports_dict = sorted(sas_ports_dict.items(),
+                                    key=lambda kv: int(kv[0].split(':')[1]))
+            health = []
+            if sas_ports_dict:
+                for rid, rate in sas_ports_dict:
+                    r_data["health"]["specifics"].append({
+                        "resource_id": f"{sas_rid}:{rid.strip()}",
+                        "negotiated_link_rate": rate.strip()
+                    })
+                    if "Gbit".lower() in rate.strip().lower():
+                        health.append("OK")
+
+            # By some logic set health status, description, recommendation
+            health_data.append(health)
+            health_status = "OK" if "OK" in health else "Fault"
+            self.set_health_data(r_data, health_status)
+            sas_hba_data.append(r_data)
+
+        if leaf_node == "sas_hba":
+            return sas_hba_data
+
+        sas_ports_data = []
+        for heath_list in health_data:
+            for i, health in enumerate(heath_list):
+                if i % 4:
+                    continue
+
+                sas_port_data = self.get_data_template(f'sas_port_{i/4+1}',
+                                                       False)
+                sas_port_data["health"]["specifics"].append({
+                    "negotiated_link_rate": "12.0 Gbit"
+                })
```

Comment:
Hi @satish-seagate , I have hard coded this for now, but for sas_port we are combining 4 phy links and referring to it as sas_port_1 and again declaring it's health as "OK" if any one of the 4 links has "OK" status, so what should be the link_rate for this sas_port??
Should it be value of any one of them or average of all four??

Reply:
if all PHYs are in good health, then SAS port health is good. No port level link rate can be given i think(dont have detailed info here, need to dig). So in specifics, list all PHYs, respective link rate  & its status probably

Reply:
need to change the schema in that case.

Reply:
Schema is for knowing mandatory keys and basic data under specifics. Here while implementing we need to fill specifics with more relevant data